File: setup/src/cabal/webapi/routers/actions.py
"""Action-safety HTTP surface: prepare, execute, and the mutation audit listing."""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter, Body, Depends, Request

from cabal.webapi import security
from cabal.webapi.envelope import ApiError, envelope_response

# Auth is declared at router level so routes keep the guard when flattened onto the app.
router = APIRouter(dependencies=[Depends(security.require_bearer_token)])


@router.post("/api/actions/{action_id}/prepare")
def prepare_action(request: Request, action_id: str, params: dict[str, Any] = Body(default={})):
    ticket = request.app.state.actions.prepare(action_id, params, request.app.state)
    return envelope_response(
        data=ticket,
        source="actions",
        precondition_digest=ticket.get("precondition_digest"),
    )


@router.post("/api/actions/{action_id}/execute")
def execute_action(request: Request, action_id: str, payload: dict[str, Any] = Body(...)):
    ticket_id = payload.get("ticket_id")
    if not isinstance(ticket_id, str) or not ticket_id:
        raise ApiError(422, "params_invalid", "execute body requires a string 'ticket_id'")
    http_status, data = request.app.state.actions.execute(action_id, ticket_id, request.app.state)
    return envelope_response(data=data, source="actions", http_status=http_status)


@router.get("/api/audit")
def list_audit(request: Request):
    return envelope_response(data={"entries": request.app.state.audit.list()}, source="audit")

File: apps/cabal-desktop/src/api/actions.ts
// Raw prepare/execute callers for the action-safety protocol (the only mutation path).
import { z } from "zod";
import { apiPost, requireData } from "@/api/client";
import { type ConfirmationTicket, confirmationTicketSchema } from "@/api/schemas";

const executeResultSchema = z.record(z.string(), z.unknown());

export interface ExecuteActionResult {
  jobId: string | null;
  data: Record<string, unknown> | null;
}

export async function prepareAction(
  actionId: string,
  params: Record<string, unknown>,
): Promise<ConfirmationTicket> {
  const envelope = await apiPost(
    `/api/actions/${actionId}/prepare`,
    confirmationTicketSchema,
    params,
  );
  return requireData(envelope, `prepare ${actionId}`);
}

export async function executeAction(
  actionId: string,
  ticketId: string,
): Promise<ExecuteActionResult> {
  const envelope = await apiPost(`/api/actions/${actionId}/execute`, executeResultSchema, {
    ticket_id: ticketId,
  });
  const data = requireData(envelope, `execute ${actionId}`);
  const jobIdValue = data.job_id;
  return {
    jobId: typeof jobIdValue === "string" ? jobIdValue : null,
    data,
  };
}
